pythoncode/usbTrainer.py

```diff
@@ -92,486 +93,713 @@
 #-------------------------------------------------------------------------------
-# legacyResistance2Power and vv for iMagic, using legacy protocol
-# Thanks to GoldenCheetah / GoldenCheetah / master / src / Train / Imagic.cpp
-#-------------------------------------------------------------------------------
-PowerResistanceFactor = 128866                  # TotalReverse
-legacyPowerResistanceFactor = (1 / 0.0036)      # GoldenCheetah ~= 277.778
-
-def Resistance2Power(Resistance, SpeedKmh):
-    global LegacyProtocol
+class clsTacxTrainer():
+    UsbDevice               = None          # clsHeadUnitLegacy and clsHeadUnitNew only!
+    AntDevice               = None          # clsVortexTrainer only!
+    TargetMode              = mode_Power    # Start with power mode
+    TargetGrade             = None          # no grade
+    TargetPower             = 100           # and 100Watts
+    TargetResistance        = None          # calculated and input to trainer
+
+    UserAndBikeWeight       = 75 + 10       # defined according the standard (data page 51)
+
+    Axis                    = 0
+    Buttons                 = 0
+    Cadence                 = 0             # below variables provided by trainer
+    CurrentPower            = 0
+    CurrentResistance       = 0
+    HeartRate               = 0
+    PedalEcho               = 0
+    SpeedKmh                = 0
+    TargetResistanceFT      = 0             # Returned from trainer
+    WheelSpeed              = 0
+
+    clv                     = None          # Command line variables
+    PowercurveFactor        = 1             #
+
+    # USB devices only:
+    Headunit                = 0
+    Calibrate               = 0
+    SpeedScale              = None          # To be set in sub-class
+    #---------------------------------------------------------------------------
+    # G e t T r a i n e r
+    #---------------------------------------------------------------------------
+    # Function      Create a TacxTrainer-child class, depending on the 
+    #               circumstances.
+    #
+    # Returns       Object of a TacxTrainer sub-class
+    #---------------------------------------------------------------------------
+    @staticmethod                       # pretent to be c++ factory function
+    def GetTrainer(clv, AntDevice=None):
+        if debug.on(debug.Function):logfile.Write ("GetTrainer()")
 
-    if LegacyProtocol:
-        if SpeedKmh == 0:
-            return 0
-        else:
-            # GoldenCheetah: curPower = ((curResistance * 0.0036f) + 0.2f) * curSpeedInternal;
-            # return round(((Resistance / legacyPowerResistanceFactor) + 0.2) * WheelSpeed, 1)
+        #-----------------------------------------------------------------------
+        # Usually I do not like multiple exit points, but here it's too handy
+        #-----------------------------------------------------------------------
+        if clv.SimulateTrainer: return clsSimulatedTrainer(clv)

```

Comment:
But... early exit because the remaining code is not relevant, also avoiding a long if-else with everything indented, is a style I definitely support.

Reply:
Indeed. But usually one function multiple exit points, I regret afterwards